### ISSUE TITLE
[JENKINS-73] CVS changelog always shows edited files as added

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#13406]
---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73" target="_blank">CVS changelog always shows edited files as added</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>resolution</b>: Postponed
<li><b>resolved</b>: 2006-07-28T21:40:31+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-11-24
</ul></i>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119841&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119841">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yeah, my colleages have noticied this but I haven't really paid serious
attention to this.</p>

<p>I believe CVS does offer enough information. Sounds like I need to hack the Ant
cvs changelog parser once again.</p>
</pre>
</details>

</ul>
</details>

<p>Yeah, my colleages have noticied this but I haven't really paid serious<br/>
attention to this.</p>

<p>I believe CVS does offer enough information. Sounds like I need to hack the Ant<br/>
cvs changelog parser once again.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119842&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119842">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>CVS log output looks like the following:</p>

<p>----------------------------
revision 1.16
date: 2006/07/26 21:00:10;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 147c44c7d7da4eb2;  filename: foo.txt;
changed
----------------------------
revision 1.15
date: 2006/07/26 20:57:00;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 150044c7d71c4c46;  filename: foo.txt;
changed
----------------------------
revision 1.14
date: 2006/07/26 20:55:37;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:
kv;  commitid: 166c44c7d6c84b34;  filename: foo.txt;
changed
=============================================================================</p>

<p>And '&lt;prevrevision&gt;' is obtained only when the "cvs log" command fetched more
than one entries from the server. The revision of the previous log record is
used as the 'previous revision' of the current log record.</p>

<p>Since Hudson runs "cvs log" with -d option, chances are, it only selects log
records that happened since the last build, and in that case, changelog.xml
doesn't record &lt;prevrevision&gt; because CVS didn't return any.</p>

<p>Therefore, it seems like after all it's not always possible to determine
add/edit/remove just from the CVS log command. We can distinguish remove from
other two by looking for "state: dead" line, but we can't distinguish add and
edit unless we check the previous revision.</p>

<p>I thought about using the output from "cvs update", but I still don't think we
can distinguish add and edit. I believe both may produce 'U' record.</p>

<p>Perhaps doing this correctly depends on issue #49. Then we can reliably find out
what data we are exchanging with the CVS server.</p>
</pre>
</details>

</ul>
</details>

<p>CVS log output looks like the following:</p>

<p>----------------------------<br/>
revision 1.16<br/>
date: 2006/07/26 21:00:10;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 147c44c7d7da4eb2;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.15<br/>
date: 2006/07/26 20:57:00;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 150044c7d71c4c46;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.14<br/>
date: 2006/07/26 20:55:37;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 166c44c7d6c84b34;  filename: foo.txt;<br/>
changed<br/>
=============================================================================</p>

<p>And '<prevrevision>' is obtained only when the "cvs log" command fetched more<br/>
than one entries from the server. The revision of the previous log record is<br/>
used as the 'previous revision' of the current log record.</p>

<p>Since Hudson runs "cvs log" with -d option, chances are, it only selects log<br/>
records that happened since the last build, and in that case, changelog.xml<br/>
doesn't record <prevrevision> because CVS didn't return any.</p>

<p>Therefore, it seems like after all it's not always possible to determine<br/>
add/edit/remove just from the CVS log command. We can distinguish remove from<br/>
other two by looking for "state: dead" line, but we can't distinguish add and<br/>
edit unless we check the previous revision.</p>

<p>I thought about using the output from "cvs update", but I still don't think we<br/>
can distinguish add and edit. I believe both may produce 'U' record.</p>

<p>Perhaps doing this correctly depends on issue #49. Then we can reliably find out<br/>
what data we are exchanging with the CVS server.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119843&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119843">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Meanwhile, I improved the parser to at least detect deletion.</p>
</pre>
</details>

</ul>
</details>

<p>Meanwhile, I improved the parser to at least detect deletion.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119844&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119844">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Add vs. Edit is tricky because you can have a file 1.1 deleted in 1.2 and
readded in 1.3, so just checking rev.equals("1.1") will not always work. Also,
adding on a branch creates 1.1 as DEAD and a real rev 1.1.${2*k}.1 for some k&gt;0,
and I presume readding a dead file on a branch produces some other form still.
So without a full file history it is hard to do more than guess. Since Hudson
now only asks the server for logs for files which really changed in the last
update, maybe it could leave off the -d arg to cvs log and try to extract the
important bits itself. Ugh.</p>

<p>Anyway I consider this a relatively low priority issue; it took me a long time
to even notice that document_add.gif was supposed to mean "added file". The file
name and log message are the important bits - if you really want to know what
changed you had better go find the diff, e.g. subscribe to the CVS change list
or have a ViewCVS server installed somewhere.</p>
</pre>
</details>

</ul>
</details>

<p>Add vs. Edit is tricky because you can have a file 1.1 deleted in 1.2 and<br/>
readded in 1.3, so just checking rev.equals("1.1") will not always work. Also,<br/>
adding on a branch creates 1.1 as DEAD and a real rev 1.1.${2*k}.1 for some k>0,<br/>
and I presume readding a dead file on a branch produces some other form still.<br/>
So without a full file history it is hard to do more than guess. Since Hudson<br/>
now only asks the server for logs for files which really changed in the last<br/>
update, maybe it could leave off the -d arg to cvs log and try to extract the<br/>
important bits itself. Ugh.</p>

<p>Anyway I consider this a relatively low priority issue; it took me a long time<br/>
to even notice that document_add.gif was supposed to mean "added file". The file<br/>
name and log message are the important bits - if you really want to know what<br/>
changed you had better go find the diff, e.g. subscribe to the CVS change list<br/>
or have a ViewCVS server installed somewhere.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-73?focusedId=119845&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-119845">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I fixed "1.1" handling. If we know "1.1" is not remove, then it's always add,
because it cannot be edit.</p>

<p>The rest has to wait for 49. Like you say, file names and comments are the
important bits.</p>
</pre>
</details>

</ul>
</details>

<p>I fixed "1.1" handling. If we know "1.1" is not remove, then it's always add,<br/>
because it cannot be edit.</p>

<p>The rest has to wait for 49. Like you say, file names and comments are the<br/>
important bits.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `depends_on` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-49%22">JENKINS-49</a>]</i>
